File: OTCDerivativesCalculatorModule/Project_CSharp/XmlBuilder/XmlSerializedClass/fpml-shared-5-4/FxSpotRateSource.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Xml;

namespace FpmlSerializedCSharp
{
    public class FxSpotRateSource : ISerialized
    {
        public FxSpotRateSource(XmlNode xmlNode)
        {
            XmlNode primaryRateSourceNode = xmlNode.SelectSingleNode("primaryRateSource");
            
            if (primaryRateSourceNode != null)
            {
                if (primaryRateSourceNode.Attributes["href"] != null || primaryRateSourceNode.Attributes["id"] != null) 
                {
                    if (primaryRateSourceNode.Attributes["id"] != null) 
                    {
                        primaryRateSourceIDRef_ = primaryRateSourceNode.Attributes["id"].Value;
                        InformationSource ob = new InformationSource(primaryRateSourceNode);
                        IDManager.SetID(primaryRateSourceIDRef_, ob);
                    }
                    else if (primaryRateSourceNode.Attributes["href"] != null)
                    {
                        primaryRateSourceIDRef_ = primaryRateSourceNode.Attributes["href"].Value;
                    }
                    else
                    {
                        primaryRateSource_ = new InformationSource(primaryRateSourceNode);
                    }
                }
                else
                {
                    primaryRateSource_ = new InformationSource(primaryRateSourceNode);
                }
            }
            
        
            XmlNode secondaryRateSourceNode = xmlNode.SelectSingleNode("secondaryRateSource");
            
            if (secondaryRateSourceNode != null)
            {
                if (secondaryRateSourceNode.Attributes["href"] != null || secondaryRateSourceNode.Attributes["id"] != null) 
                {
                    if (secondaryRateSourceNode.Attributes["id"] != null) 
                    {
                        secondaryRateSourceIDRef_ = secondaryRateSourceNode.Attributes["id"].Value;
                        InformationSource ob = new InformationSource(secondaryRateSourceNode);
                        IDManager.SetID(secondaryRateSourceIDRef_, ob);
                    }
                    else if (secondaryRateSourceNode.Attributes["href"] != null)
                    {
                        secondaryRateSourceIDRef_ = secondaryRateSourceNode.Attributes["href"].Value;
                    }
                    else
                    {
                        secondaryRateSource_ = new InformationSource(secondaryRateSourceNode);
                    }
                }
                else
                {
                    secondaryRateSource_ = new InformationSource(secondaryRateSourceNode);
                }
            }
            
        
            XmlNode fixingTimeNode = xmlNode.SelectSingleNode("fixingTime");
            
            if (fixingTimeNode != null)
            {
                if (fixingTimeNode.Attributes["href"] != null || fixingTimeNode.Attributes["id"] != null) 
                {
                    if (fixingTimeNode.Attributes["id"] != null) 
                    {
                        fixingTimeIDRef_ = fixingTimeNode.Attributes["id"].Value;
                        BusinessCenterTime ob = new BusinessCenterTime(fixingTimeNode);
                        IDManager.SetID(fixingTimeIDRef_, ob);
                    }
                    else if (fixingTimeNode.Attributes["href"] != null)
                    {
                        fixingTimeIDRef_ = fixingTimeNode.Attributes["href"].Value;
                    }
                    else
                    {
                        fixingTime_ = new BusinessCenterTime(fixingTimeNode);
                    }
                }
                else
                {
                    fixingTime_ = new BusinessCenterTime(fixingTimeNode);
                }
            }
            
        
        }
        
    
        #region PrimaryRateSource_
        private InformationSource primaryRateSource_;
        public InformationSource PrimaryRateSource_
        {
            get
            {
                if (this.primaryRateSource_ != null)
                {
                    return this.primaryRateSource_; 
                }
                else if (this.primaryRateSourceIDRef_ != null)
                {
                    primaryRateSource_ = IDManager.getID(primaryRateSourceIDRef_) as InformationSource;
                    return this.primaryRateSource_; 
                }
                else
                {
                      return this.primaryRateSource_; 
                }
            }
            set
            {
                if (this.primaryRateSource_ != value)
                {
                    this.primaryRateSource_ = value;
                }
            }
        }
        #endregion
        
        public string primaryRateSourceIDRef_ { get; set; }
        #region SecondaryRateSource_
        private InformationSource secondaryRateSource_;
        public InformationSource SecondaryRateSource_
        {
            get
            {
                if (this.secondaryRateSource_ != null)
                {
                    return this.secondaryRateSource_; 
                }
                else if (this.secondaryRateSourceIDRef_ != null)
                {
                    secondaryRateSource_ = IDManager.getID(secondaryRateSourceIDRef_) as InformationSource;
                    return this.secondaryRateSource_; 
                }
                else
                {
                      return this.secondaryRateSource_; 
                }
            }
            set
            {
                if (this.secondaryRateSource_ != value)
                {
                    this.secondaryRateSource_ = value;
                }
            }
        }
        #endregion
        
        public string secondaryRateSourceIDRef_ { get; set; }
        #region FixingTime_
        private BusinessCenterTime fixingTime_;
        public BusinessCenterTime FixingTime_
        {
            get
            {
                if (this.fixingTime_ != null)
                {
                    return this.fixingTime_; 
                }
                else if (this.fixingTimeIDRef_ != null)
                {
                    fixingTime_ = IDManager.getID(fixingTimeIDRef_) as BusinessCenterTime;
                    return this.fixingTime_; 
                }
                else
                {
                      return this.fixingTime_; 
                }
            }
            set
            {
                if (this.fixingTime_ != value)
                {
                    this.fixingTime_ = value;
                }
            }
        }
        #endregion
        
        public string fixingTimeIDRef_ { get; set; }
        
    
        
    
    }
    
}

File: OTCDerivativesCalculatorModule/Project_Cpp/lib_static/FpmlSerialized/GenClass/fpml-shared-5-4/FxSpotRateSource.cpp
// FxSpotRateSource.cpp 
#include "FxSpotRateSource.hpp"
#ifdef ConsolePrint
    #include <iostream>
#endif 

namespace FpmlSerialized {

FxSpotRateSource::FxSpotRateSource(TiXmlNode* xmlNode)
: ISerialized(xmlNode)
{
    #ifdef ConsolePrint
        std::string initialtap_ = FileManager::instance().tap_;
        FileManager::instance().tap_.append("   ");
    #endif 
   //primaryRateSourceNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* primaryRateSourceNode = xmlNode->FirstChildElement("primaryRateSource");

   if(primaryRateSourceNode){primaryRateSourceIsNull_ = false;}
   else{primaryRateSourceIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- primaryRateSourceNode , address : " << primaryRateSourceNode << std::endl;
   #endif
   if(primaryRateSourceNode)
   {
      if(primaryRateSourceNode->Attribute("href") || primaryRateSourceNode->Attribute("id"))
      {
          if(primaryRateSourceNode->Attribute("id"))
          {
             primaryRateSourceIDRef_ = primaryRateSourceNode->Attribute("id");
             primaryRateSource_ = boost::shared_ptr<InformationSource>(new InformationSource(primaryRateSourceNode));
             primaryRateSource_->setID(primaryRateSourceIDRef_);
             IDManager::instance().setID(primaryRateSourceIDRef_,primaryRateSource_);
          }
          else if(primaryRateSourceNode->Attribute("href")) { primaryRateSourceIDRef_ = primaryRateSourceNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { primaryRateSource_ = boost::shared_ptr<InformationSource>(new InformationSource(primaryRateSourceNode));}
   }

   //secondaryRateSourceNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* secondaryRateSourceNode = xmlNode->FirstChildElement("secondaryRateSource");

   if(secondaryRateSourceNode){secondaryRateSourceIsNull_ = false;}
   else{secondaryRateSourceIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- secondaryRateSourceNode , address : " << secondaryRateSourceNode << std::endl;
   #endif
   if(secondaryRateSourceNode)
   {
      if(secondaryRateSourceNode->Attribute("href") || secondaryRateSourceNode->Attribute("id"))
      {
          if(secondaryRateSourceNode->Attribute("id"))
          {
             secondaryRateSourceIDRef_ = secondaryRateSourceNode->Attribute("id");
             secondaryRateSource_ = boost::shared_ptr<InformationSource>(new InformationSource(secondaryRateSourceNode));
             secondaryRateSource_->setID(secondaryRateSourceIDRef_);
             IDManager::instance().setID(secondaryRateSourceIDRef_,secondaryRateSource_);
          }
          else if(secondaryRateSourceNode->Attribute("href")) { secondaryRateSourceIDRef_ = secondaryRateSourceNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { secondaryRateSource_ = boost::shared_ptr<InformationSource>(new InformationSource(secondaryRateSourceNode));}
   }

   //fixingTimeNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* fixingTimeNode = xmlNode->FirstChildElement("fixingTime");

   if(fixingTimeNode){fixingTimeIsNull_ = false;}
   else{fixingTimeIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- fixingTimeNode , address : " << fixingTimeNode << std::endl;
   #endif
   if(fixingTimeNode)
   {
      if(fixingTimeNode->Attribute("href") || fixingTimeNode->Attribute("id"))
      {
          if(fixingTimeNode->Attribute("id"))
          {
             fixingTimeIDRef_ = fixingTimeNode->Attribute("id");
             fixingTime_ = boost::shared_ptr<BusinessCenterTime>(new BusinessCenterTime(fixingTimeNode));
             fixingTime_->setID(fixingTimeIDRef_);
             IDManager::instance().setID(fixingTimeIDRef_,fixingTime_);
          }
          else if(fixingTimeNode->Attribute("href")) { fixingTimeIDRef_ = fixingTimeNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { fixingTime_ = boost::shared_ptr<BusinessCenterTime>(new BusinessCenterTime(fixingTimeNode));}
   }

    #ifdef ConsolePrint
        FileManager::instance().tap_ = initialtap_;
    #endif 
}
boost::shared_ptr<InformationSource> FxSpotRateSource::getPrimaryRateSource()
{
   if(!this->primaryRateSourceIsNull_){
        if(primaryRateSourceIDRef_ != ""){
             return boost::shared_static_cast<InformationSource>(IDManager::instance().getID(primaryRateSourceIDRef_));
        }else{
             return this->primaryRateSource_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<InformationSource>();
   }
}
boost::shared_ptr<InformationSource> FxSpotRateSource::getSecondaryRateSource()
{
   if(!this->secondaryRateSourceIsNull_){
        if(secondaryRateSourceIDRef_ != ""){
             return boost::shared_static_cast<InformationSource>(IDManager::instance().getID(secondaryRateSourceIDRef_));
        }else{
             return this->secondaryRateSource_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<InformationSource>();
   }
}
boost::shared_ptr<BusinessCenterTime> FxSpotRateSource::getFixingTime()
{
   if(!this->fixingTimeIsNull_){
        if(fixingTimeIDRef_ != ""){
             return boost::shared_static_cast<BusinessCenterTime>(IDManager::instance().getID(fixingTimeIDRef_));
        }else{
             return this->fixingTime_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<BusinessCenterTime>();
   }
}
}

